internal/pkg/orchestrator/service.go

```diff
@@ -0,0 +1,209 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package orchestrator
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+
+	"go.opentelemetry.io/auto/internal/pkg/instrumentors"
+	"go.opentelemetry.io/auto/internal/pkg/process"
+)
+
+// Service is responsible for managing all instrumentation.
+type Service struct {
+	ctx         context.Context
+	version     string
+	analyzer    *process.Analyzer
+	exePath     string
+	serviceName string
+	monitorAll  bool
+	processch   chan *pidServiceName
+	deadProcess chan int
+	managers    map[int]*instrumentors.Manager
+	exporter    sdktrace.SpanExporter
+	pid         int
+	pidTicker   <-chan time.Time
+}
+
+type ServiceOpt interface {
+	apply(Service) Service
+}
+
+const (
+	// envTargetExeKey is the key for the environment variable value pointing to the
+	// target binary to instrument.
+	envTargetExeKey = "OTEL_GO_AUTO_TARGET_EXE"
+	// envServiceName is the key for the envoriment variable value containing the service name.
+	envServiceNameKey = "OTEL_SERVICE_NAME"
+	// envResourceAttrKey is the key for the environment variable value containing
+	// OpenTelemetry Resource attributes.
+	envResourceAttrKey = "OTEL_RESOURCE_ATTRIBUTES"
+	// serviceNameDefault is the default service name prefix used if a user does not provide one.
+	serviceNameDefault = "unknown_service"
+)
+
+func (s Service) applyEnv() Service {
+	if v, ok := os.LookupEnv(envTargetExeKey); ok {
+		s.exePath = v
+	}
+	if v, ok := os.LookupEnv(envServiceNameKey); ok {
+		s.serviceName = v
+	} else {
+		s = s.applyResourceAtrrEnv()
+		if s.serviceName == "" {
+			s = s.setDefualtServiceName()
+		}
+	}
+	return s
+}
+
+func (s Service) setDefualtServiceName() Service {
+	if s.exePath != "" {
+		s.serviceName = fmt.Sprintf("%s:%s", serviceNameDefault, filepath.Base(s.exePath))
+	} else {
+		s.serviceName = serviceNameDefault
+	}
+	return s
+}
+
+func (s Service) applyResourceAtrrEnv() Service {
+	attrs := strings.TrimSpace(os.Getenv(envResourceAttrKey))
+	serviceName := serviceNameFromAttrs(attrs)
+	if serviceName != "" {
+		s.serviceName = serviceName
+	}
+	return s
+}
+
+func serviceNameFromAttrs(attrs string) string {
+	serviceName := ""
+	if attrs == "" {
+		return serviceName
+	}
+
+	pairs := strings.Split(attrs, ",")
+	for _, p := range pairs {
+		k, v, found := strings.Cut(p, "=")
+		if !found {
+			continue
+		}
+		key := strings.TrimSpace(k)
+		if key == string(semconv.ServiceNameKey) {
+			serviceName = strings.TrimSpace(v)
+		}
+	}
+	return serviceName
+}
+
+func (s Service) Validate() error {
+	if s.pid != 0 {
+		return validatePID(s.pid)
+	}
+
+	if s.monitorAll {
+		return nil
+	}
+	if s.serviceName == "" {
+		return fmt.Errorf("serviceName is nil")
+	}
+
+	if s.exePath == "" {
+		return fmt.Errorf("execPath is nil")
+	}
```

Comment:
This will only evaluate at most one of these conditionals. Is this intentional?

Reply:
Yes. It is the current behaviour as well.